### ISSUE TITLE
array_walk with pass-by-reference causes false positive unused variable

### ISSUE THREAD

**Issue by sirbrillig** [sirbrillig/phpcs-variable-analysis#215]
This code produces `Unused variable $value`, when it is in fact used:

```php
     array_walk($userNameParts, static function (string &$value): void {
            if (strlen($value) <= 3) {
                return;
            }

            $value = ucfirst($value);
        });
```

Reported by @Levivb in https://github.com/sirbrillig/phpcs-variable-analysis/issues/214

**Comment by sirbrillig** [comment]
This is supposed to be handled by the code here: https://github.com/sirbrillig/phpcs-variable-analysis/blob/29b6bb402db30777503a2687c849456ed4f741be/VariableAnalysis/Sniffs/CodeAnalysis/VariableAnalysisSniff.php#L1675

**Comment by sirbrillig** [comment]
Actually, it looks like the variable is correctly being marked as pass-by-reference, but then when it reaches where the variable is used the code thinks that the variable is a new instance; more specifically, that it has a new scope. The original variable is defined as having a scope starting with the array_walk closure, but the new variable's scope starts at the enclosing function.

The bug is happening because the scope detection code is getting confused about the difference between a bound reference variable and one which is going to be bound at runtime.

For example:

```php
$foo = 'hello';
$bar = &$foo; // $bar is now a reference variable bound to $foo
$bar = 'bye'; // this writes to the bound variable; now $foo is 'bye'
```

```php
array_walk($data, function (&$foo) { // $foo is now a reference variable that will be bound at runtime
  $foo = ucfirst($foo); // this writes to the bound variable, but we can't know anything about that variable since it does not exist in the code before runtime
});
```
